benches/opencl.rs

```diff
@@ -0,0 +1,65 @@
+use criterion::{criterion_group, criterion_main, BatchSize, Criterion};
+use rand::Rng;
+use std::time::{Duration, Instant};
+use num_format::{Locale, ToFormattedString};
+
+fn random_bytes<const BYTES: usize>() -> Vec<u8> {
+    let mut bytes = vec![0u8; BYTES];
+    rand::thread_rng().fill(&mut bytes[..]);
+    bytes
+}
+
+fn random_bytes_inplace<const BYTES: usize>(vec: &mut Vec<u8>) {
+    rand::thread_rng().fill(&mut vec[..]);
+}
+
+pub fn criterion_benchmark(c: &mut Criterion) {
+    let mut group = c.benchmark_group("opencl");
+    group.sample_size(100);
+    group.measurement_time(Duration::from_secs(30));
+
+    const NUM_PIECES: usize = 1024 * 1024; // * 2 + 1024 * 56;
+    const MAX_LAYERS: usize = 1;
+    let size: f64 = NUM_PIECES as f64 * 4096 as f64 / 1073741824 as f64;
+    let expanded_ivs = random_bytes::<{32 * NUM_PIECES}>();
+
+    let instances = sloth256_189::opencl::initialize().unwrap();
+
+    let mut layers: usize = 1;
+    while layers <= MAX_LAYERS {
+        group.bench_function(format!("Encode-parallel/{} GB/{} pieces/{} layer(s)", size, NUM_PIECES.to_formatted_string(&Locale::en), layers), |b| {
+            let piece = random_bytes::<{4096 * NUM_PIECES}>();
+
+            b.iter_batched_ref(
+                || piece.clone(),
+                |mut input| sloth256_189::opencl::encode(&mut input, &expanded_ivs, layers, instances).unwrap(),

```

Comment:
Is there a good reason to not use pinned memory at all?

Reply:
Found this in: https://forums.developer.nvidia.com/t/advantages-disadvantages-of-using-pinned-memory/34422

> The only real drawbacks to pinning memory are the reduction in available physical ram to the host demand-paging system, and the time it takes to pin memory (which is significantly longer than an ordinary malloc of the same size).

Also, I remember that pinned memory can't be pre-empted.

Reply:
So whether it should be the `default` for our case, I would like to hear your opinions as well @simonatsn :)

Reply:
The main reason to consider not using pinned memory is based on some limitations in OpenCL. Nvidia's support for OpenCL limits how much space you can allocate for pinned memory (~1/3 of GPU memory) and it allocates twice that much on the GPU at runtime. So half of that is wasted. Other than that, for performance reasons I don't think there is any reason you would not want to use it. The approach taken in this change is to allocate the buffers once so the additional overhead there is negligible.